server/src/main/java/org/zanata/magpie/model/TextFlowMTRequest.java

```diff
@@ -0,0 +1,131 @@
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.zanata.magpie.model;
+
+import java.io.Serializable;
+import java.util.Date;
+import java.util.List;
+
+import javax.persistence.Access;
+import javax.persistence.AccessType;
+import javax.persistence.Column;
+import javax.persistence.ElementCollection;
+import javax.persistence.Entity;
+import javax.persistence.EnumType;
+import javax.persistence.Enumerated;
+import javax.persistence.FetchType;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Temporal;
+import javax.persistence.TemporalType;
+import javax.validation.constraints.NotNull;
+
+import org.hibernate.annotations.NaturalId;
+import com.google.common.collect.Lists;
+
+/**
+ * @author Patrick Huang
+ * <a href="mailto:pahuang@redhat.com">pahuang@redhat.com</a>
+ */
+@Entity
+@Access(AccessType.FIELD)
+public class TextFlowMTRequest implements Serializable {
+    private static final long serialVersionUID = -6874010575058871413L;
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Temporal(TemporalType.TIMESTAMP)
+    @Column(nullable = false)
+    @NotNull
+    private Date invokeDate;
+
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    @NotNull
+    private BackendID backendID;
+
+    @NaturalId
+    @ManyToOne
+    @JoinColumn(name = "document_id", updatable = false, nullable = false)
+    @NotNull
+    private Document document;
+
+    @NaturalId
+    @ManyToOne(optional = false)
+    @JoinColumn(name = "locale_id", nullable = false, updatable = false)
+    @NotNull
+    private Locale locale;
+
+    @ManyToOne(optional = false)
+    @JoinColumn(name = "triggered_account_id", nullable = false, updatable = false)
+    @NotNull
+    private Account triggeredBy;
+
+    @ElementCollection(fetch = FetchType.EAGER)
+    private List<String> textFlowContentHashes = Lists.newArrayList();
```

Comment:
@huangp Is this the new table you meant? I'm having a bit of trouble working out how we would reconcile a Google Translator bill against this information.

Could you please check my assumptions here?

1. In this context, "Document" refers to eg an article, ie many text flows. From the Document, you can get to the full set of text flows and their exact contents.
2. But it's entirely possible that only a **subset** of the Document's text flows are part of a given translation request.
3. The column textflowcontenthashes is limited to 128 chars
4. There seems to be no foreign key for textflowcontenthashes against the TextFlow table.

Is textflowcontenthashes really just **one** content hash, despite the plural name? If it's meant to hold multiple hashes, 128 seems too short.

If we ever needed to use the data stored in TextFlowMTRequest and textflowcontenthashes, how would we work out what text we sent off for translation, or how much of it there was? It doesn't seem like the content hashes are enough to find the corresponding TextFlows easily, nor do we store the character count.

It might be better (and simpler) if TextFlowMTRequest just held a string count, word count and character count, without worrying about the individual TextFlows. Then it should be straightforward to get a total of the translations which were requested in a given month (for example) with a simple SQL query.